File: src/main/java/ru/hollowhorizon/hc/fabric/internal/NetworkHelper.java
//? if fabric {
package ru.hollowhorizon.hc.fabric.internal;

import kotlin.Unit;
import net.minecraft.client.Minecraft;
import ru.hollowhorizon.hc.common.utils.JavaHacks;
import ru.hollowhorizon.hc.common.network.HollowPacketKt;

public class NetworkHelper {
    public static void register() {
        HollowPacketKt.registerPacket = (type) -> {
            FabricNetworkKt.registerPacket(JavaHacks.forceCast(type));
            return Unit.INSTANCE;
        };
        HollowPacketKt.sendPacketToClient = (player, hollowPacketV3) -> {
            player.connection.send(HollowPacketKt.asVanillaPacket(hollowPacketV3, true));
            return Unit.INSTANCE;
        };
        HollowPacketKt.sendPacketToServer = (hollowPacketV3) -> {
            var connection = Minecraft.getInstance().getConnection();
            if (connection != null) connection.send(HollowPacketKt.asVanillaPacket(hollowPacketV3, false));
            return Unit.INSTANCE;
        };
        HollowPacketKt.registerPackets.invoke();
    }
}
//?}

File: src/main/java/ru/hollowhorizon/hollowengine/fabric/internal/NetworkHelper.kt
//? if fabric {
package ru.hollowhorizon.hollowengine.fabric.internal

import kotlinx.coroutines.launch
import kotlinx.coroutines.yield
import net.minecraft.client.Minecraft
import net.minecraft.server.level.ServerPlayer
import ru.hollowhorizon.hollowengine.HollowEngine
import ru.hollowhorizon.hollowengine.common.coroutines.coroutineScope
import ru.hollowhorizon.hollowengine.common.network.*
import ru.hollowhorizon.hollowengine.common.utils.JavaHacks

object NetworkHelper {
    fun register() {
        registerPacket = { type: Class<*> ->
            registerPacket<HollowPacket>(JavaHacks.forceCast(type))
        }
        sendPacketToClient = { player: ServerPlayer, hollowPacketV3: HollowPacket ->
            player.server.coroutineScope.launch {
                while(player.connection == null && !player.isRemoved) {
                    yield() // Если пакет отправляется до инициализации, то ждём каждый тик
                }
                if (player.connection != null && !player.isRemoved) {
                    player.connection.send(hollowPacketV3.asVanillaPacket(true))
                } else {
                    HollowEngine.LOGGER.warn("Player ${player.name} removed, but packet still trying to send")
                }
            }
        }
        sendPacketToServer = { hollowPacketV3: HollowPacket? ->
            val connection = Minecraft.getInstance().getConnection()
            if (connection != null) connection.send(hollowPacketV3!!.asVanillaPacket(false))
            Unit
        }
        registerPackets.invoke()
    }
} //?}
